xspec/xspec
# Make tests independent of working environment

This pull request derives from #287. So needs to be handled after that.

This change is related only to `[XSpec]/test/`, Travis and AppVeyor. Nothing is changed in the other files.

---

Problems:

* The current tests expect `SAXON_CP` to be a single Saxon jar file. But, in your working environment, `SAXON_CP` may be [multiple classpath entries](https://github.com/xspec/xspec/wiki/XML-Catalog-support#using-an-xml-catalog-with-xspecbat-and-xspecsh).
* In your working environment, `SAXON_CP` may contain custom class. Using it for testing is not good.
* In your working environment, `SAXON_HOME` may be predefined. That may affect testing.
* The naming convention (`*_CP`) is not clear about whether it is classpath or jar.

This pull request

* uses `*_JAR` to specify Saxon and other jar files for testing.
* resets `SAXON_CP` and `SAXON_HOME` in the test scripts.
* makes the test scripts check some prerequisites and tell what's missing.

As a result,

* The tests are less affected by your working environment.
* If you test XSpec on your local machine, it's now easier to figure out what you forgot to set up.


### How to test

Here's what I tested. (This may also help #117 a bit.)

1. Uninstall Ant or remove Ant from `PATH`. (to emulate a fresh environment)

1. Open terminal and go to `[XSpec]/test/`.

1. Run `run-bats.sh` or `.cmd`.
    * You should see an error that Ant is not found.

1. Install Ant or include Ant in `PATH`.

1. Unset these environment variables. (to emulate a fresh environment)
    ```
    unset BASEX_JAR
    unset JING_JAR
    unset SAXON_JAR
    unset XML_RESOLVER_JAR
    unset XMLCALABASH_JAR
    ```
    
    Windows
    ```
    set BASEX_JAR=
    set JING_JAR=
    set SAXON_JAR=
    set XML_RESOLVER_JAR=
    set XMLCALABASH_JAR=
    ```

1. Set bogus environment variables. (to ensure that they do not affect the testing environment)
    ```
    export SAXON_CP=bogus
    export SAXON_HOME=bogus
    ```
    
    Windows
    ```
    set SAXON_CP=bogus
    set SAXON_HOME=bogus
    ```

1. Run `run-bats.sh` or `.cmd` again.
    * You should see an error that `SAXON_JAR` is not found.

1. Set `SAXON_JAR` to specify your Saxon jar file.
    ```
    export SAXON_JAR=/path/to/saxon.jar
    ```
    
    ```
    set SAXON_JAR=\path\to\saxon.jar
    ```

1. Run `run-bats.sh` or `.cmd` again.
    * Now the tests should run.
    * Some tests may be skipped. The message should tell you which environment variable is missing. (such as BASEX_JAR)
    * Set the missing environment variable to specify an appropriate jar file.
    * Repeat this step until no tests are skipped.

1. Windows: Make sure `SAXON_CP` and `SAXON_HOME` are not polluted and are still `bogus`.
    ```
    C:\test>set SAXON_
    SAXON_CP=bogus
    SAXON_HOME=bogus
    SAXON_JAR=\path\to\saxon.jar
    ```

1. Run `run-xspec-tests-ant.sh` or `.cmd`.
    * It should run XSpec unit tests and exit with `BUILD SUCCESSFUL`.

1. Go to `[XSpec]/test/end-to-end/`.

1. `run-e2e-tests.sh` or `.cmd`.
    * It should run XSpec end-to-end tests and print `OK` lines.

1. Windows: Make sure `SAXON_CP` and `SAXON_HOME` are not polluted and are still `bogus`.

1. Run `generate-expected.sh` or `.cmd`
    * It should run XSpec tests and exit gracefully.

1. Windows: Make sure `SAXON_CP` and `SAXON_HOME` are not polluted and are still `bogus`.

1. Push the commit. Verify that the tests are successful or skipped appropriately on [Travis](https://travis-ci.org/AirQuick/xspec/builds/417153800) and [AppVeyor](https://ci.appveyor.com/project/AirQuick/xspec/build/631-clarify-cp-and-jar).

1. Push a temporary commit (95a28b65fa8eb3c4f8f216f3562e24a7154c0450) which injects an error into `xspec.sh` and `xspec.bat`. Verify that both [Travis](https://travis-ci.org/AirQuick/xspec/builds/417162630) and [AppVeyor](https://ci.appveyor.com/project/AirQuick/xspec/build/632-clarify-cp-and-jar) fail.
